flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/TimeBoundedStreamInnerJoin.scala

```diff
@@ -0,0 +1,533 @@
+      backPressureSuggestion =
+        if (leftRecord) (oppositeWatermark + allowedLateness - cleanUpTime)
+        else -(oppositeWatermark + allowedLateness - cleanUpTime)
+      LOG.warn("The clean timer for the " +
+        s"${if (leftRecord) "left" else "right"}" +
+        s" stream is lower than ${if (leftRecord) "right" else "left"} watermark." +
+        s" requiredTime = ${formatTime(cleanUpTime)}, watermark = ${formatTime(oppositeWatermark)},"
+        + s"backPressureSuggestion = " + s"${backPressureSuggestion}.")
+    }
+  }
+
+
+  /**
+    * Called when a registered timer is fired.
+    * Remove records which are earlier than the expiration time,
+    * and register a new timer for the earliest remaining records.
+    *
+    * @param timestamp the timestamp of the timer
+    * @param ctx       the context to register timer or get current time
+    * @param out       the collector for returning result values
+    */
+  override def onTimer(
+    timestamp: Long,
+    ctx: CoProcessFunction[CRow, CRow, CRow]#OnTimerContext,
+    out: Collector[CRow]): Unit = {
+    getCurrentOperatorTime(ctx)
+    //TODO In the future, we should separate the left and right watermarks. Otherwise, the
+    //TODO registered timer of the faster stream will be delayed, even if the watermarks have
+    //TODO already been emitted by the source.
+    if (leftTimerState.value == timestamp) {
+      val rightExpirationTime = leftOperatorTime - rightRelativeSize - allowedLateness - 1
+      removeExpiredRecords(
+        timestamp,
+        rightExpirationTime,
+        leftOperatorTime,
+        rightCache,
+        leftTimerState,
+        ctx,
+        false
+      )
+    }
+
+    if (rightTimerState.value == timestamp) {
+      val leftExpirationTime = rightOperatorTime - leftRelativeSize - allowedLateness - 1
+      removeExpiredRecords(
+        timestamp,
+        leftExpirationTime,
+        rightOperatorTime,
+        leftCache,
+        rightTimerState,
+        ctx,
+        true
+      )
+    }
+  }
+
+  /**
+    * Remove the expired records. Register a new timer if the cache still holds records
+    * after the cleaning up.
+    */
+  private def removeExpiredRecords(
+    timerFiringTime: Long,
+    expirationTime: Long,
+    oppositeWatermark: Long,
+    recordCache: MapState[Long, JList[Row]],
+    timerState: ValueState[Long],
+    ctx: CoProcessFunction[CRow, CRow, CRow]#OnTimerContext,
+    removeLeft: Boolean): Unit = {
+
+    val keysIterator = recordCache.keys().iterator()
+
+    // Search for expired timestamps.
+    // If we find a non-expired timestamp, remember the timestamp and leave the loop.
+    // This way we find all expired timestamps if they are sorted without doing a full pass.
+    var earliestTimestamp: Long = -1L
+    var recordTime: Long = 0L
+    while (keysIterator.hasNext) {
+      //TODO The "short-circuit" code was commented, because when using a StateMap with
```

Comment:
I think you are right, short-cutting doesn't work. The worst case would be if the keys are in descending timestamp order which might happen if there are no order guarantees.